plonk-hashing/src/poseidon/poseidon_ref.rs

```diff
@@ -0,0 +1,525 @@
+    ) {
+        for (element, round_constant) in state
+            .iter_mut()
+            .zip(constants.round_constants.iter().skip(*constants_offset))
+        {
+            // element.com_addi(c, round_constant);
+            *element = Self::addi(c, element, round_constant)
+        }
+
+        *constants_offset += WIDTH;
+    }
+
+    fn product_mds(
+        c: &mut COM,
+        constants: &PoseidonConstants<Self::ParameterField>,
+        state: &mut [Self::Field; WIDTH],
+    ) {
+        let matrix = &constants.mds_matrices.m;
+        let mut result = Self::zeros::<WIDTH>(c);
+        for (j, val) in result.iter_mut().enumerate() {
+            for (i, row) in matrix.iter_rows().enumerate() {
+                // *val += row[j] * state[i];
+                let tmp = Self::muli(c, &state[i], &row[j]);
+                *val = Self::add(c, val, &tmp);
+            }
+        }
+        *state = result;
+    }
+
+    /// return (x + pre_add)^5 + post_add
+    fn quintic_s_box(
+        c: &mut COM,
+        x: Self::Field,
+        pre_add: Option<Self::ParameterField>,
+        post_add: Option<Self::ParameterField>,
+    ) -> Self::Field {
+        let mut tmp = match pre_add {
+            Some(a) => Self::addi(c, &x, &a),
+            None => x.clone(),
+        };
+        tmp = Self::power_of_5(c, &tmp);
+        match post_add {
+            Some(a) => Self::addi(c, &tmp, &a),
+            None => tmp,
+        }
+    }
+
+    fn power_of_5(c: &mut COM, x: &Self::Field) -> Self::Field {
+        let mut tmp = Self::mul(c, x, x); // x^2
+        tmp = Self::mul(c, &tmp, &tmp); // x^4
+        Self::mul(c, &tmp, x) // x^5
+    }
+
+    fn alloc(c: &mut COM, v: Self::ParameterField) -> Self::Field;
+    fn zeros<const W: usize>(c: &mut COM) -> [Self::Field; W];
+    fn zero(c: &mut COM) -> Self::Field {
+        Self::zeros::<1>(c)[0].clone()
+    }
+    fn add(c: &mut COM, x: &Self::Field, y: &Self::Field) -> Self::Field;
+    fn addi(
+        c: &mut COM,
+        a: &Self::Field,
+        b: &Self::ParameterField,
+    ) -> Self::Field;
+    fn mul(c: &mut COM, x: &Self::Field, y: &Self::Field) -> Self::Field;
+    fn muli(
+        c: &mut COM,
+        x: &Self::Field,
+        y: &Self::ParameterField,
+    ) -> Self::Field;
+}
+
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct PoseidonRef<COM, S: PoseidonRefSpec<COM, WIDTH>, const WIDTH: usize>
+where
+    S: ?Sized,
+{
+    pub(crate) constants_offset: usize,
+    pub(crate) current_round: usize,
+    pub elements: [S::Field; WIDTH],
+    pos: usize,
+    pub(crate) constants: PoseidonConstants<S::ParameterField>,
+}
+
+impl<COM, S: PoseidonRefSpec<COM, WIDTH>, const WIDTH: usize>
+    PoseidonRef<COM, S, WIDTH>
+{
+    pub fn new(
+        c: &mut COM,
+        constants: PoseidonConstants<S::ParameterField>,
+    ) -> Self {
+        let mut elements = S::zeros(c);
+        elements[0] = S::alloc(c, constants.domain_tag);
```

Comment:
This loads `constants.domain_tag` into a witness variable using `c.add_input`, but how do we constrain the witness to correspond to the actual constants?

Reply:
Good point... a naive way is that for the input phase we can just use an `zero_var` for `element[0]` and keep track of domain tag, and then we can handle this in the first round of `add_round_constants` and s-box.